esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5878,55 +5878,53 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
     return iRet;
 }
 
-unsigned CWsDfuEx::getFilePartsInfo(IEspContext &context, IDistributedFile *df, const char *clusterName,
-    IArrayOf<IEspDFUPartLocations> &dfuPartLocations, IArrayOf<IEspDFUPartCopies> &dfuPartCopies)
+void CWsDfuEx::getFilePartsInfo(IEspContext &context, Owned<IFileDescriptor> &fdesc, IArrayOf<IEspDFUFilePart> &dfuParts)
 {
-    int nextLocationsIndex = 1; // NB: both LocationIndex and PartIndex are 1 based in response.
-    MapStringTo<int> locationMap;
-    Owned<IFileDescriptor> fdesc = df->getFileDescriptor(clusterName);
+    // NB: both CopyIndex and PartIndex are 1 based in response.
     Owned<IPartDescriptorIterator> pi = fdesc->getIterator();
     ForEach(*pi)
     {
         IPartDescriptor& part = pi->query();
-        unsigned partIndex = part.queryPartIndex();
 
-        StringArray partCopyLocationsIndexes;
+        IArrayOf<IEspDFUFileCopy> fileCopies;
         for (unsigned int i=0; i<part.numCopies(); i++)
         {
-            StringBuffer host, locationsIndexStr;
+            StringBuffer host, path;
             part.queryNode(i)->endpoint().getUrlStr(host);
-            int *locationsIndex = locationMap.getValue(host.str());
-            if (locationsIndex)
-            {
-                partCopyLocationsIndexes.append(locationsIndexStr.append(*locationsIndex).str());
-                continue;
-            }
-
-            Owned<IEspDFUPartLocations> partLocations = createDFUPartLocations();
-            partLocations->setLocationIndex(nextLocationsIndex);
-            partLocations->setLocation(host.str());
-            dfuPartLocations.append(*partLocations.getClear());
+            part.getPath(path, i);
 
-            partCopyLocationsIndexes.append(locationsIndexStr.append(nextLocationsIndex).str());
-            locationMap.setValue(host.str(), nextLocationsIndex);
-            nextLocationsIndex++;
+            Owned<IEspDFUFileCopy> fileCopy = createDFUFileCopy();
+            fileCopy->setCopyIndex(i + 1);
+            fileCopy->setHost(host.str());
+            fileCopy->setPath(path.str());
```

Comment:
I can separate the getFilePartsInfo() to 2 copies, one for FileAcesss and one the FileCreate. Or I can switch back to old FilePartLocations and FileParts structure. Please let me know what you want.

Reply:
The FileAccess response should remain as it was FilePartLocations and FileParts.
The new structure for FileCreate is fine, and necessary for now, because we are bypassing dafilesrv, but eventually it will likely change to be similar to the existing FileAccess response.

Reply:
It's okay to have the optional Path member of DFUFileCopy, but it should only be filled in for DFUFileCreateResponse, and left blank otherwise.